[spec] Update syntax spec for memory64

sbc100 opened the pull request:
This is just a start.  I've made some updates to `instructions.rst` and `types.rst` but could use some pointers so uploading now to get feedback.

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -189,6 +190,8 @@ Memory Types
 .. math::
    \begin{array}{llll}
+   \production{index type} & \indextype &::=&
+     \u32 ~|~ \u64 \\
```
These would be values. I think you mean `\I32` and `\I64` for the actual types.

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -547,8 +547,10 @@ Instructions in this group are concerned with linear :ref:`memory <syntax-mem>`.
 
 .. math::
    \begin{array}{llcl}
+   \production{index type} & \X{it} &::=&
```
```suggestion
   \production{index value} & \X{iv} &::=&
```
However, the grammar for u64 subsumes the one for u32, so you could just as well use u64 below. The binary grammar only uses u64 anyways. So we can just drop this.

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -583,19 +585,21 @@ Instructions in this group are concerned with linear :ref:`memory <syntax-mem>`.
 Memory is accessed with |LOAD| and |STORE| instructions for the different :ref:`number types <syntax-numtype>`.
 They all take a *memory immediate* |memarg| that contains an address *offset* and the expected *alignment* (expressed as the exponent of a power of 2).
+The type of the *offset* corresponds to the *index type* of the memory which will
+be either |U32| (for 32-bit memories) or |U64| (for 64-bit memories).
```
I don't think that's correct. It will be decoded/parsed as u64 in all cases (because parser/decoder don't know the types), but there is a range check in the validator.

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -583,19 +585,21 @@ Instructions in this group are concerned with linear :ref:`memory <syntax-mem>`.
 Alternatively, vector loads can perform a *splat*, such that only a single lane of the specified storage size is loaded, and the result is duplicated to all lanes.
 
-The static address offset is added to the dynamic address operand, yielding a 33 bit *effective address* that is the zero-based index at which the memory is accessed.
+The static address offset is added to the dynamic address operand.  The type of
+the dynamic, like the static offset, corresponds to the *index type* of
+memory.  The result of this addition is a 33 bit or 65 bit *effective address*
```
?

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -189,6 +190,8 @@ Memory Types
+   \production{index type} & \indextype &::=&
+     \u32 ~|~ \u64 \\
    \production{memory type} & \memtype &::=&
      \limits \\
```
```suggestion
     \indextype~\limits \\
```

sbc100 commented on document/core/syntax/types.rst:
```diff
@@ -190,7 +193,7 @@ Memory Types
    \begin{array}{llll}
    \production{memory type} & \memtype &::=&
-     \limits \\
+     \idxtype~\limits.\X{it} \\
```
I think I need to say that the idxtype matches the `it` suffix on the limits here, but I'm not sure how to do that.

rossberg replied:
By introducing a subscript on limits, as in the Overview.

rossberg replied:
On second thought, the more adequate notation would be parameters in parentheses, like we already use for [vectors](https://webassembly.github.io/spec/core/syntax/conventions.html#vectors):
```
limits(iv) ::= {min iv, max iv?}

memtype ::= idxtype limits(type(iv))
```
I'll update the overview with this.

rossberg reviewed (commented):
FWIW, you can generally copy what's in the Overview pretty directly.

sbc100 commented:
This PR is now just missing validation and execution changes.

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -156,6 +156,18 @@ They are also used to classify the inputs and outputs of :ref:`instructions <syn
+
+Index Type
+~~~~~~~~~~
+
```
There should be some explanatory sentence here, like with the other sections.

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -190,7 +202,15 @@ Memory Types
    \begin{array}{llll}
    \production{memory type} & \memtype &::=&
-     \limits \\
+     ~\idxtype~\limits(type(\X{iv})) \\
```
I think you mean something like `\limits(\F{val}(\idxtype))`, since you'd need to map the idxtype to a value representation.

However, I don't think this dependency works. In the binary grammar, decoding of limits can produce 64-bit limits, independent of the index type. So the abstract syntax must allow this. Instead of constraining the value range in the AST, it needs to be plain i64, and a range check has to occur in validation.

(In theory, we could make the binary decoder dependent on the index type as well, to match this, but such a type dependency is quite alien to how Wasm works elsewhere.)

rossberg replied:
Ah, scratch that, I see now that the intention is that the dependency is already built in into the decoder (which I missed, see latest comment).

Okay, so this would work, but it requires dependent types to work, so at least the reference interpreter cannot express it accurately. Since it does not add anything useful, I think it's still better to avoid this extra complexity.

sbc100 replied:
Are you saying we don't need/want to be parameterize `limits` like this?   I believe this was introduced to the overview by you in b52270d6f93291d2bf0ca9820d904aa7ed4dc5f8

rossberg replied:
Yes, sorry, my bad, it wasn't a good idea (I actually introduced it long ago in #13; that commit was just adjusting the notation — and screwed up the invocation of the `type` function on the way).

So yes, I think we should change it back. Now that tables are also 64-bit, we can even simplify it further relative to what it was before #13 and drop the subscript/parameter on limits altogether.

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -169,8 +181,8 @@ Limits
    \begin{array}{llll}
-   \production{limits} & \limits &::=&
-     \{ \LMIN~\u32, \LMAX~\u32^? \} \\
+   \production{limits} & \limits(iv) &::=&
+     \{ \LMIN~\X{iv}, \LMAX~\X{iv}^? \} \\
```
```suggestion
   \production{limits} & \limits &::=&
     \{ \LMIN~\I64, \LMAX~\I64^? \} \\
```
(see next comment)

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -211,7 +231,15 @@ Table Types
    \begin{array}{llll}
    \production{table type} & \tabletype &::=&
-     \limits~\reftype \\
+     ~\idxtype~\limits(type(\X{iv})) ~\reftype \\
```
```suggestion
     ~\idxtype~\limits ~\reftype \\
```

rossberg commented on document/core/text/modules.rst:
```diff
@@ -361,7 +361,7 @@ A :ref:`data segment <text-data>` can be given inline with a memory definition,
    \production{module field} &
-     \text{(}~\text{memory}~~\Tid^?~~\text{(}~\text{data}~~b^n{:}\Tdatastring~\text{)}~~\text{)} \quad\equiv \\ & \qquad
+     \text{(}~\text{memory}~~\Tid^?~~\Tidxtype^?~~\text{(}~\text{data}~~b^n{:}\Tdatastring~\text{)}~~\text{)} \quad\equiv \\ & \qquad
        \text{(}~\text{memory}~~\Tid'~~m~~m~\text{)} \\ & \qquad
```
```suggestion
       \text{(}~\text{memory}~~\Tid'≈\Tidxtype^?~~m~~m~\text{)} \\ & \qquad
```
Also, the same change is needed for the corresponding abbreviation for tables.

rossberg commented on document/core/text/types.rst:
```diff
@@ -139,7 +152,8 @@ Memory Types
    \begin{array}{llclll@{\qquad\qquad}l}
    \production{memory type} & \Tmemtype &::=&
-     \X{lim}{:}\Tlimits &\Rightarrow& \X{lim} \\
+     \X{lim}{:}\Tlimits &\Rightarrow& \I32~\X{lim} \\ &&|&
```
This should be defined separately now, like other abbreviations.

sbc100 replied:
Done

rossberg commented on document/core/text/types.rst:
```diff
@@ -153,7 +167,8 @@ Table Types
    \begin{array}{llclll}
    \production{table type} & \Ttabletype &::=&
-     \X{lim}{:}\Tlimits~~\X{et}{:}\Treftype &\Rightarrow& \X{lim}~\X{et} \\
+     \X{lim}{:}\Tlimits~~\X{et}{:}\Treftype &\Rightarrow& \I32~\X{lim}~\X{et} \\ &&|&
```
Same here.

rossberg commented on document/core/binary/types.rst:
```diff
@@ -130,7 +130,9 @@ Limits
    \begin{array}{llclll}
    \production{limits} & \Blimits &::=&
      \hex{00}~~n{:}\Bu32 &\Rightarrow& \{ \LMIN~n, \LMAX~\epsilon \} \\ &&|&
-     \hex{01}~~n{:}\Bu32~~m{:}\Bu32 &\Rightarrow& \{ \LMIN~n, \LMAX~m \} \\
+     \hex{01}~~n{:}\Bu32~~m{:}\Bu32 &\Rightarrow& \{ \LMIN~n, \LMAX~m \} \\ &&|&
+     \hex{04}~~n{:}\Bu64 &\Rightarrow& \{ \LMIN~n, \LMAX~\epsilon \} \\ &&|&
+     \hex{05}~~n{:}\Bu64~~m{:}\Bu64 &\Rightarrow& \{ \LMIN~n, \LMAX~m \}
```
Shouldn't all these return a pair of idxtype and limits?

rossberg replied:
Also, the decoding width of the values depends on the index type bit. This does not match what's currently implemented in the interpreter, which always uses the same. Is there any benefit in introducing this dependency?

(To clarify, I suggest using Bu64 in all cases, which would still be backwards-compatible. That's what's currently implemented in the interpreter.)

sbc100 replied:
Done

sbc100 commented:
All comments addressed.

The Execution and Validation section are still not complete.  I updated these for just `memory.copy` to see if I'm on the right track?    Once I get memory.copy right I'll update the rest.

sbc100 commented on document/core/exec/instructions.rst:
```diff
@@ -2351,65 +2351,67 @@ Memory Instructions
+15. If :math:`d \leq s`, then:
 
-   a. Push the value :math:`\I32.\CONST~d` to the stack.
+   a. Push the value :math:`\X{it}.\CONST~d` to the stack.
```
I'm not sure if I can use `it` directly here or if I have to map it somehow?

rossberg replied:
That's totally fine. But note that we now have multiple memories, so there will be two different it's.

sbc100 commented on document/core/valid/instructions.rst:
```diff
@@ -1186,13 +1186,15 @@ Memory Instructions
+     C.\CMEMS[0] = it~\limits
    }{
-     C \vdashinstr \MEMORYCOPY : [\I32~\I32~\I32] \to []
+     C \vdashinstr \MEMORYCOPY : [it~it~it] \to []
```
Again, my main question is how/if I need to map the index type here?

rossberg replied:
Same here.

sbc100 commented on document/core/valid/instructions.rst:
```diff
@@ -1186,13 +1186,15 @@ Memory Instructions
 * The memory :math:`C.\CMEMS[0]` must be defined in the context.
 
-* Then the instruction is valid with type :math:`[\I32~\I32~\I32] \to []`.
+* Let :math:`it~\limits` be the :ref:`memory type <syntax-memtype>` :math:`C.\CMEMS[0]`.
```
Should use idxtype here here rather than the `it` abbreviation?

rossberg replied:
it is fine, just make sure to wrap it in `\X{...}` to have proper Latex type setting.

rossberg commented on document/core/binary/types.rst:
```diff
@@ -124,13 +124,15 @@ Function Types
    \production{limits} & \Blimits &::=&
-     \hex{00}~~n{:}\Bu32 &\Rightarrow& \{ \LMIN~n, \LMAX~\epsilon \} \\ &&|&
-     \hex{01}~~n{:}\Bu32~~m{:}\Bu32 &\Rightarrow& \{ \LMIN~n, \LMAX~m \} \\
+     \hex{00}~~n{:}\Bu64 &\Rightarrow& \I32~~\{ \LMIN~n, \LMAX~\epsilon \} \\ &&|&
+     \hex{01}~~n{:}\Bu64~~m{:}\Bu64 &\Rightarrow& \I32~~\{ \LMIN~n, \LMAX~m \} \\ &&|&
+     \hex{04}~~n{:}\Bu64 &\Rightarrow& \I64~~\{ \LMIN~n, \LMAX~\epsilon \} \\ &&|&
+     \hex{05}~~n{:}\Bu64~~m{:}\Bu64 &\Rightarrow& \I64~~\{ \LMIN~n, \LMAX~m \}
```
To match the use site in the rules for memtype and tabletype, you wanna return these as pairs (it, lim).

sbc100 replied:
Sorry do you mean use braces around the RHS and the comma between the two values?   I don't see other productions here doing that so maybe I'm misunderstanding?

rossberg replied:
At least a comma, since this is a meta-level pair. There are a couple of places where this occurs, see e.g. the text format rule for [`elemlist`](https://webassembly.github.io/spec/core/text/modules.html#text-elemlist).

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -156,6 +156,21 @@ They are also used to classify the inputs and outputs of :ref:`instructions <syn
+Index Type
+~~~~~~~~~~
+
+Index types classify the type of the values that can be used to index into
+:ref:`memories <syntax-mem>` and :ref:`tables <syntax-table>`.
```
```suggestion
*Index types* classify the values that can be used to index into :ref:`memories <syntax-mem>` or :ref:`tables <syntax-table>`.
```

rossberg commented on document/core/text/modules.rst:
```diff
@@ -361,11 +361,13 @@ A :ref:`data segment <text-data>` can be given inline with a memory definition,
-     \text{(}~\text{memory}~~\Tid^?~~\text{(}~\text{data}~~b^n{:}\Tdatastring~\text{)}~~\text{)} \quad\equiv \\ & \qquad
-       \text{(}~\text{memory}~~\Tid'~~m~~m~\text{)} \\ & \qquad
-       \text{(}~\text{data}~~\text{(}~\text{memory}~~\Tid'~\text{)}~~\text{(}~\text{i32.const}~~\text{0}~\text{)}~~\Tdatastring~\text{)}
+     \text{(}~\text{memory}~~\Tid^?~~it^?~~\text{(}~\text{data}~~b^n{:}\Tdatastring~\text{)}~~\text{)} \quad\equiv \\ & \qquad
```
```suggestion
     \text{(}~\text{memory}~~\Tid^?~~\X{it}^?~~\text{(}~\text{data}~~b^n{:}\Tdatastring~\text{)}~~\text{)} \quad\equiv \\ & \qquad
```

rossberg commented on document/core/text/modules.rst:
```diff
@@ -361,11 +361,13 @@ A :ref:`data segment <text-data>` can be given inline with a memory definition,
-       \text{(}~\text{memory}~~\Tid'~~m~~m~\text{)} \\ & \qquad
-       \text{(}~\text{data}~~\text{(}~\text{memory}~~\Tid'~\text{)}~~\text{(}~\text{i32.const}~~\text{0}~\text{)}~~\Tdatastring~\text{)}
+     \text{(}~\text{memory}~~\Tid^?~~it^?~~\text{(}~\text{data}~~b^n{:}\Tdatastring~\text{)}~~\text{)} \quad\equiv \\ & \qquad
+       \text{(}~\text{memory}~~\Tid'~~it'~~m~~m~\text{)} \\ & \qquad
```
```suggestion
       \text{(}~\text{memory}~~\Tid'~~\X{it}^?~~m~~m~\text{)} \\ & \qquad
```

rossberg commented on document/core/text/modules.rst:
```diff
@@ -361,11 +361,13 @@ A :ref:`data segment <text-data>` can be given inline with a memory definition,
-       \text{(}~\text{data}~~\text{(}~\text{memory}~~\Tid'~\text{)}~~\text{(}~\text{i32.const}~~\text{0}~\text{)}~~\Tdatastring~\text{)}
+     \text{(}~\text{memory}~~\Tid^?~~it^?~~\text{(}~\text{data}~~b^n{:}\Tdatastring~\text{)}~~\text{)} \quad\equiv \\ & \qquad
+       \text{(}~\text{memory}~~\Tid'~~it'~~m~~m~\text{)} \\ & \qquad
+       \text{(}~\text{data}~~\text{(}~\text{memory}~~\Tid'~\text{)}~~\text{(}~it'~\text{.const}~~\text{0}~\text{)}~~\Tdatastring~\text{)}
```
```suggestion
       \text{(}~\text{data}~~\text{(}~\text{memory}~~\Tid'~\text{)}~~\text{(}~\X{it}'\text{.const}~~\text{0}~\text{)}~~\Tdatastring~\text{)}
```

rossberg commented on document/core/text/modules.rst:
```diff
@@ -361,11 +361,13 @@ A :ref:`data segment <text-data>` can be given inline with a memory definition,
        \\ & \qquad\qquad
-       (\iff \Tid^? \neq \epsilon \wedge \Tid' = \Tid^? \vee \Tid^? = \epsilon \wedge \Tid' \idfresh, m = \F{ceil}(n / 64\,\F{Ki})) \\
+       (\iff \Tid^? \neq \epsilon \wedge \Tid' = \Tid^? \vee \Tid^? = \epsilon \wedge \Tid' \idfresh, \\ & \qquad\qquad
+        \iff it^? \neq \epsilon \wedge it' = it^? \vee it^? = \epsilon \wedge it' = \text{i32}, \\ & \qquad\qquad
```
```suggestion
        \land \x{it}^? \neq \epsilon \wedge \X{it}' = \X{it}^? \vee \X{it}^? = \epsilon \wedge \X{it}' = \text{i32} \\ & \qquad\qquad
```

rossberg commented on document/core/text/modules.rst:
```diff
@@ -361,11 +361,13 @@ A :ref:`data segment <text-data>` can be given inline with a memory definition,
-       (\iff \Tid^? \neq \epsilon \wedge \Tid' = \Tid^? \vee \Tid^? = \epsilon \wedge \Tid' \idfresh, m = \F{ceil}(n / 64\,\F{Ki})) \\
+       (\iff \Tid^? \neq \epsilon \wedge \Tid' = \Tid^? \vee \Tid^? = \epsilon \wedge \Tid' \idfresh, \\ & \qquad\qquad
+        \iff it^? \neq \epsilon \wedge it' = it^? \vee it^? = \epsilon \wedge it' = \text{i32}, \\ & \qquad\qquad
+        m = \F{ceil}(n / 64\,\F{Ki})), \\
```
```suggestion
       \land m = \F{ceil}(n / 64\,\F{Ki})), \\
```

rossberg commented on document/core/text/types.rst:
```diff
@@ -139,7 +154,16 @@ Memory Types
+
+Abbreviations
+.............
+
```
Add a short sentence here explaining the short-hand.

tlively replied:
Since abbreviations are supposed to be in-place expansions, doesn't this production need to be part of the main index type syntax rather than being an abbreviation?

sbc100 replied:
Are you saying I've not expressed the abbreviation correctly, or is there some reason why this type of abbreviation can't be expressed as an in-place expansion?

What I'm trying to say is specify just the `limits` expands to the `"i32" limits`.

tlively replied:
This can't be an in-place expansion because it would replace _every_ `limits` with `'i32' limits`. For example, if the text module contained `... i64 0 0 ...` it would expand to `... i64 i32 0 0 ...`. The expansion would never terminate and it would also produce an invalid module.

sbc100 replied:
I see, that makes sense there is nothing anchor the expansion.

rossberg replied:
I think the expansion here should be one for index types, defined as `ε ≡ I32`.

tlively replied:
What kind of contexts
 are the expansions evaluated in? I thought we couldn’t do expansions in specific contexts like when we know an index type is expected.

rossberg replied:
That is left a bit vague, intended to be applicable to a specific syntactic class (which currently only appears in the Latex source :) ) and otherwise "clear from context". As mentioned in my SpecTec presentation last week, I'd like to remove the handwaviness and make this more precise, but that doesn't have super-high priority.

tlively replied:
Oh wow, that wasn't my interpretation at all. I thought you could have arbitrary `(param)` sprinkled in any and all locations throughout your wat and it would be ok because it would expand out to nothing before parsing. I'm glad to hear that's not the intention, though.

rossberg commented on document/core/text/types.rst:
```diff
@@ -153,7 +177,16 @@ Table Types
+
+Abbreviations
+.............
+
```
Same here.

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -156,6 +156,21 @@ They are also used to classify the inputs and outputs of :ref:`instructions <syn
    \end{array}
 
 
+.. _syntax-idxtype:
```
```suggestion
.. _index:: ! index type
   pair: abstract syntax; index type
   single: memory; index type
   single: table; index type
.. _syntax-idxtype:
```

rossberg reviewed (commented):
Some more suggestions for tweaks.

Before continuing to spell out execution and validation, I'd merge the multi-memory repo into this one, so that you don't have to redo everything later (with merge conflicts everywhere). (And please merge it without squashing commits, otherwise we'll have to resolve possible conflicts over each time.) I'd also suggest doing all that in a separate PR.

sbc100 commented:
> Some more suggestions for tweaks.
> 
> Before continuing to spell out execution and validation, I'd merge the multi-memory repo into this one, so that you don't have to redo everything later (with merge conflicts everywhere). (And please merge it without squashing commits, otherwise we'll have to resolve possible conflicts over each time.) I'd also suggest doing all that in a separate PR.

Sure thing.  However the multi-memory repo itself looks like its fairly out of date with upstream, resulting is a fair amount of conflicts.  Would you be able update the multi-memory repo first?  (I tried myself and there were a fair few conflicts.. let me know if you would rather I try to take care of that too).

rossberg commented:
Ah sorry. Instead of the multi-memory repo, you could merge the wasm-3.0 branch from the main repo, which already contains it, plus various other proposals (it is the staging ground for merging phase-4 proposals and is up-to-date with main).

sbc100 commented:
> Ah sorry. Instead of the multi-memory repo, you could merge the wasm-3.0 branch from the main repo, which already contains it, plus various other proposals (it is the staging ground for merging phase-4 proposals and is up-to-date with main).

I'm having a little trouble with the wasm-3.0 branch merge for reasons that look unrelated to memoy64.

I think the reason is because i'm already caught up with the latest `main` changes but wasm-3.0 doesn't have all changes from main (e.g. https://github.com/WebAssembly/spec/pull/1751).     Would it be possible to merge the latest changes from main into wasm-3.0, or should I rebase memory64 on top branch point of wasm-3.0 perhaps?

rossberg commented:
Done, merged spec/main into wasm-3.0.

sbc100 commented:
I'm most of the way through the rebase onto wasm-3.0.

Would you be OK with land this change here on main and then I can cherry pick it over to my wasm-3.0 branch once that lands?

Any other feedback on the content of this PR?

rossberg commented on document/core/text/types.rst:
```diff
@@ -139,7 +154,16 @@ Memory Types
+
+.. math::
+   \begin{array}{llclll@{\qquad\qquad}l}
+   \production{memory type} & \Tmemtype &::=&
+     \X{lim}{:}\Tlimits &\Rightarrow& \I32~\X{lim} \\
```
```suggestion
   \production{index type} &
     \epsilon &\equiv& \I32 \\
```
This abbreviation should rather go to the section on index type (and then nothing needs to be said here or in the table type section).

sbc100 commented:
The merge of main with upstream/wasm-3.0 is now complete.  See https://github.com/WebAssembly/memory64/tree/wasm-3.0.

This PR should now be on top of memory64/wasm3.0.   Should I close this one and open another one, or first push and change the target branch to wasm-3.0?

sbc100 committed:
[spec] Update syntax spec for memory64

I updated `instructions.rst` and make a start on `types.rst` but could
use some help on how to parameterize `limits` on the index type.

rossberg commented:
I'd avoid reopening a PR if possible. If you can rebase it, that seems highly preferable.

sbc100 commented:
This PR has now been updated to target the local wasm-3.0 branch.

The validation and execution sections still only contain updates for `memory.copy`.  I will do the rest once we finish iterating on those.

rossberg commented on document/core/binary/instructions.rst:
```diff
@@ -278,7 +278,7 @@ Each variant of :ref:`memory instruction <syntax-instr-memory>` is encoded with
-     a{:}\Bu32~~o{:}\Bu32 &\Rightarrow& 0~\{ \ALIGN~a,~\OFFSET~o \}
+     a{:}\Bu32~~o{:}\Bu64 &\Rightarrow& 0~\{ \ALIGN~a,~\OFFSET~o \}
        & (\iff a < 2^6) \\ &&|&
      a{:}\Bu32~~x{:}\memidx~~o{:}\Bu32 &\Rightarrow& x~\{ \ALIGN~(a - 2^6),~\OFFSET~o \}
```
```suggestion
     a{:}\Bu32~~x{:}\memidx~~o{:}\Bu64 &\Rightarrow& x~\{ \ALIGN~(a - 2^6),~\OFFSET~o \}
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3682,65 +3682,71 @@ Memory Instructions
+11. Let :math:`\X{it}_s~\limits` be the :math:`\X{mem}_s.\MITYPE`.
 
-12. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+12. Let :math:`\X{it}_{min}` be ....?
```
```suggestion
12. Let :math:`\X{it}_{\X{min}}` be ....?
```
And similarly in other places. Or perhaps name this one just plain "it"?

I would say "... be the :ref:`minimum <aux-idxtype-min>` of ...` and then define that as a convention in the syntax section for index types.

sbc100 replied:
Done.  Although I'm not sure how you would me to define `aux-idxtype-min` itself.   Right now I've described it in prose but I imagine it warrants some kind of formal notation.

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3682,65 +3682,71 @@ Memory Instructions
+21. If :math:`d \leq s`, then:
 
-   a. Push the value :math:`\I32.\CONST~d` to the stack.
+   a. Push the value :math:`\X{it}.\CONST~d` to the stack.
```
There is no plain it defined above, this will need to be one of it_d, it_s, or it_min. Likewise below.

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -349,6 +349,25 @@ In a :ref:`module <syntax-module>`, each member of a recursive type is assigned
+   single: table; index type
+.. _syntax-idxtype:
+
+Index Type
```
I have come to wonder whether "index type" is the appropriate name, since we use "index" for static indices only. Perhaps "offset type" (offtype) would be more appropriate. But we can figure that out in a later PR.

sbc100 replied:
Fair enough.  I'll open another issue for that: https://github.com/WebAssembly/memory64/issues/67

rossberg commented on document/core/text/modules.rst:
```diff
@@ -365,11 +365,13 @@ A :ref:`data segment <text-data>` can be given inline with a memory definition,
-       \text{(}~\text{data}~~\text{(}~\text{memory}~~\Tid'~\text{)}~~\text{(}~\text{i32.const}~~\text{0}~\text{)}~~\Tdatastring~\text{)}
+     \text{(}~\text{memory}~~\Tid^?~~\X{it}^?~~\text{(}~\text{data}~~b^n{:}\Tdatastring~\text{)}~~\text{)} \quad\equiv \\ & \qquad
+       \text{(}~\text{memory}~~\Tid'~~\X{it}^?~~m~~m~\text{)} \\ & \qquad
+       \text{(}~\text{data}~~\text{(}~\text{memory}~~\Tid'~\text{)}~~\text{(}~\X{it}'~\text{.const}~~\text{0}~\text{)}~~\Tdatastring~\text{)}
```
```suggestion
       \text{(}~\text{data}~~\text{(}~\text{memory}~~\Tid'~\text{)}~~\text{(}~\X{it}'\text{.const}~~\text{0}~\text{)}~~\Tdatastring~\text{)}
```

rossberg commented on document/core/valid/instructions.rst:
```diff
@@ -1753,15 +1753,17 @@ Memory Instructions
 * The memory :math:`C.\CMEMS[y]` must be defined in the context.
 
-* Then the instruction is valid with type :math:`[\I32~\I32~\I32] \to []`.
+* Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`C.\CMEMS[0]`.
```
With multi memory, the index type can't come from mems[0], but needs to be the minimum of mems[x] and mems[y].

rossberg commented on document/core/valid/instructions.rst:
```diff
@@ -1753,15 +1753,17 @@ Memory Instructions
 .. math::
    \frac{
-     C.\CMEMS[x] = \memtype
+     C.\CMEMS[x] = \X{it1}~\limits1
```
```suggestion
     C.\CMEMS[x] = \X{it}_1~\limits_1
```

rossberg commented on document/core/valid/instructions.rst:
```diff
@@ -1753,15 +1753,17 @@ Memory Instructions
+     C.\CMEMS[x] = \X{it1}~\limits1
      \qquad
-     C.\CMEMS[x] = \memtype
+     C.\CMEMS[y] = \X{it2}~\limits2
```
```suggestion
     C.\CMEMS[y] = \X{it}_2~\limits_2
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3682,65 +3682,71 @@ Memory Instructions
+10. Let :math:`\X{it}_d~\limits` be the :math:`\X{mem}_d.\MITYPE`.
 
-11. Pop the value :math:`\I32.\CONST~n` from the stack.
+11. Let :math:`\X{it}_s~\limits` be the :math:`\X{mem}_s.\MITYPE`.
```
This could well have different limits than the other memory, so you need to subscript \limits, too.

sbc100 committed:
feedback

sbc100 committed:
validation rules

sbc100 commented:
In addition to addressing feedback I also updated all he instruction validation rules. 

AFAIK, that just leaves the completion of the execution section (in addition to any remaining feedback).

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3710,41 +3710,41 @@ Memory Instructions
+   h. Push the value :math:`\X{it}_s.\CONST~s` to the stack.
 
-23. Push the value :math:`\X{it}_{min}.\CONST~(n-1)` to the stack.
+23. Push the value :math:`\X{it}.\CONST~(n-1)` to the stack.
```
```suggestion
23. Push the value :math:`\X{it}_n.\CONST~(n-1)` to the stack.
```

rossberg commented on document/core/syntax/types.rst:
```diff
@@ -367,6 +367,10 @@ Index Type
      \I32 ~|~ \I64 \\
    \end{array}
 
+.. _aux-idxtype-min:
+
+The minimum of two index types is defined as |I32| if either of the types are
+|I32| and |I64| otherwise.
```
```suggestion
.. _aux-idxtype-min:

Convention
..........

The *minimum* of two index types is defined as |I32| if either of the types are |I32|,
and |I64| otherwise.

.. math::
   \begin{array}{llll}
   \itmin(\I64, \I64) &=& \I64 \\
   \itmin(\X{it}_1, \X{it}_2) &=& \I32 & (\otherwise) \\
   \end{array}
```
Plus define itmin as a macro pointing here.

rossberg commented on document/core/valid/instructions.rst:
```diff
@@ -1753,17 +1753,21 @@ Memory Instructions
+     C.\CMEMS[y] = \X{it}_y~\limits_y
    }{
-     C \vdashinstr \MEMORYCOPY~x~y : [\X{it}~\X{it}~\X{it}] \to []
+     C \vdashinstr \MEMORYCOPY~x~y : [\X{it}_x~\X{it}_y~\X{it}] \to []
```
```suggestion
     C \vdashinstr \MEMORYCOPY~x~y : [\X{it}_x~\X{it}_y~\itmin(\X{it}_x, \X{it}_y)] \to []
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3682,65 +3682,71 @@ Memory Instructions
-21. Execute the instruction :math:`\MEMORYCOPY~x~y`.
+24. Execute the instruction :math:`\MEMORYCOPY~x~y`.
 
 .. math::
```
The formal rule here still needs the equivalent adjustments as the text.

rossberg reviewed (commented):
> In addition to addressing feedback I also updated all he instruction validation rules.

Hm, I only see an update to memory.copy. Did you perhaps forget to push?

sbc100 committed:
feedback

sbc100 committed:
exec spec

sbc100 commented:
I looked into updated the formal rules for `memory.copy` evaluation semantics.   However, the end result looks like it will be quite verbose due to repeated conditions.  @rossberg can you recommend a way to factor out the index type matching conditions?  @tlively suggested using a locally defined helper function perhaps?

rossberg commented:
The execution rules shouldn't actually need the side conditions on what the different it's are concretely. Execution does not depend on that, and we know that they match up due to validation. Leaving out the side conditions means the types exist, but we don't care what they are (here).

You can in fact do the same in the text, that is, remove the lookup of the memory types and just amend the existing type assertions on the operands to something like

"Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` :math:`\X{it}_n` is on the top of the stack, for some :ref:`index type <syntax-idxtype>` :math:`\X{it}_n`."

and so on.

sbc100 commented:
> The execution rules shouldn't actually need the side conditions on what the different it's are concretely. Execution does not depend on that, and we know that they match up due to validation. Leaving out the side conditions means the types exist, but we don't care what they are (here).
> 
> You can in fact do the same in the text, that is, remove the lookup of the memory types and just amend the existing type assertions on the operands to something like
> 
> "Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` :math:`\X{it}_n` is on the top of the stack, for some :ref:`index type <syntax-idxtype>` :math:`\X{it}_n`."
> 
> and so on.

Do what should the stack look line in the memory.copy execution rules?   Can I just refer directly to `it_x` `it_y` and `it` there?

rossberg commented:
Yes, it would be just a variable in the pattern match.

sbc100 committed:
feedback

sbc100 committed:
exec rules

sbc100 commented:
All execution rules updated.   I think that everything now % feedback.

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -2541,28 +2541,30 @@ Table Instructions
 5. Let :math:`\X{tab}` be the :ref:`table instance <syntax-tableinst>` :math:`S.\STABLES[a]`.
 
-6. Assert: due to :ref:`validation <valid-table.get>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{tab}.\TITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`table type <syntax-tabletype>` :math:`\X{tab}.\TITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -2584,30 +2586,32 @@ Table Instructions
 5. Let :math:`\X{tab}` be the :ref:`table instance <syntax-tableinst>` :math:`S.\STABLES[a]`.
 
-6. Assert: due to :ref:`validation <valid-table.set>`, a :ref:`reference value <syntax-ref>` is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{tab}.\TITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`table type <syntax-tabletype>` :math:`\X{tab}.\TITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -2629,14 +2633,16 @@ Table Instructions
 5. Let :math:`\X{tab}` be the :ref:`table instance <syntax-tableinst>` :math:`S.\STABLES[a]`.
 
-6. Let :math:`\X{sz}` be the length of :math:`\X{tab}.\TIELEM`.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{tab}.\TITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`table type <syntax-tabletype>` :math:`\X{tab}.\TITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -2658,37 +2664,39 @@ Table Instructions
 5. Let :math:`\X{tab}` be the :ref:`table instance <syntax-tableinst>` :math:`S.\STABLES[a]`.
 
-6. Let :math:`\X{sz}` be the length of :math:`S.\STABLES[a]`.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{tab}.\TITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`table type <syntax-tabletype>` :math:`\X{tab}.\TITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -2726,60 +2734,62 @@ Table Instructions
 5. Let :math:`\X{tab}` be the :ref:`table instance <syntax-tableinst>` :math:`S.\STABLES[\X{ta}]`.
 
-6. Assert: due to :ref:`validation <valid-table.fill>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{tab}.\TITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`table type <syntax-tabletype>` :math:`\X{tab}.\TITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -2809,98 +2819,104 @@ Table Instructions
 9. Let :math:`\X{tab}_y` be the :ref:`table instance <syntax-tableinst>` :math:`S.\STABLES[\X{ta}_y]`.
 
-10. Assert: due to :ref:`validation <valid-table.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+10. Let :math:`\X{it}_x~\limits_x` be the :math:`\X{tab}_x.\TITYPE`.
```
```suggestion
10. Let :math:`\X{it}_x~\limits_x` be the :ref:`table type <syntax-tabletype>` :math:`\X{tab}_x.\TITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -2809,98 +2819,104 @@ Table Instructions
+10. Let :math:`\X{it}_x~\limits_x` be the :math:`\X{tab}_x.\TITYPE`.
 
-11. Pop the value :math:`\I32.\CONST~n` from the stack.
+11. Let :math:`\X{it}_y~\limits_y` be the :math:`\X{tab}_y.\TITYPE`.
```
```suggestion
11. Let :math:`\X{it}_y~\limits_y` be the :ref:`table type <syntax-tabletype>` :math:`\X{tab}_y.\TITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -2922,76 +2938,78 @@ Table Instructions
 5. Let :math:`\X{tab}` be the :ref:`table instance <syntax-tableinst>` :math:`S.\STABLES[\X{ta}]`.
 
-6. Assert: due to :ref:`validation <valid-table.init>`, :math:`F.\AMODULE.\MIELEMS[y]` exists.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{tab}.\TITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`table type <syntax-tabletype>` :math:`\X{tab}.\TITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3057,39 +3075,41 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Assert: due to :ref:`validation <valid-loadn>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3132,33 +3152,35 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Assert: due to :ref:`validation <valid-load-extend>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3193,31 +3215,33 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Assert: due to :ref:`validation <valid-load-extend>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3250,29 +3274,31 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Assert: due to :ref:`validation <valid-load-extend>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3305,37 +3331,39 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Assert: due to :ref:`validation <valid-load-extend>`, a value of :ref:`value type <syntax-valtype>` |V128| is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3370,41 +3398,43 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Assert: due to :ref:`validation <valid-storen>`, a value of :ref:`value type <syntax-valtype>` :math:`t` is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3446,33 +3476,35 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Assert: due to :ref:`validation <valid-storen>`, a value of :ref:`value type <syntax-valtype>` :math:`\V128` is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3534,33 +3568,35 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Let :math:`\X{sz}` be the length of :math:`S.\SMEMS[a]` divided by the :ref:`page size <page-size>`.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3598,17 +3634,19 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[\X{ma}]`.
 
-6. Assert: due to :ref:`validation <valid-memory.fill>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -3808,76 +3843,78 @@ Memory Instructions
 5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[\X{ma}]`.
 
-6. Assert: due to :ref:`validation <valid-memory.init>`, :math:`F.\AMODULE.\MIDATAS[y]` exists.
+6. Let :math:`\X{it}~\limits` be the :math:`\X{mem}.\MITYPE`.
```
```suggestion
6. Let :math:`\X{it}~\limits` be the :ref:`memory type <syntax-memtype>` :math:`\X{mem}.\MITYPE`.
```

rossberg commented on document/core/valid/instructions.rst:
```diff
@@ -1505,17 +1507,19 @@ Memory Instructions
 .. math::
    \frac{
-     C.\CMEMS[x] = \memtype
+     C.\CMEMS[x] = \X{it}~\limits
```
This change is missing in all the other load/store rules, so that `it` remains undetermined there.

rossberg reviewed (commented):
LGTM % comments

sbc100 committed:
feedback
